More easily detect overridden methods & calls to super

java-coding-prodigy opened the pull request:
Fixes #982

Currently added a `JavaMethod#isOverridden` method for now.

java-coding-prodigy committed:
Added Method#isOverridden

hankem commented on archunit/src/main/java/com/tngtech/archunit/core/domain/JavaMethod.java:
```diff
@@ -125,6 +126,12 @@ public String getDescription() {
+    @PublicAPI(usage = ACCESS)
+    public boolean isOverridden() {
+        Method method = methodSupplier.get();
+        return Arrays.asList(method.getDeclaringClass().getSuperclass().getDeclaredMethods()).contains(method);
```
It would be nice to avoid the (expensive) reflection API, but more importantly: Does this even work for you? Please provide a unit test that proves the correctnes of your code.

I'd use something like this:
```java
public class JavaMethodTest {
    @Test
    public void isOverridden() {
        class Base {
            void method1() {}
            void method1(int x) {}
        }
        class Child extends Base {
            void method1() {}
            void method2() {}
        }
        JavaClass childClass = new ClassFileImporter().importClass(Child.class);
        assertThat(childClass.getMethod("method1").isOverridden()).isTrue();
        assertThat(childClass.getMethod("method1", int.class).isOverridden()).isFalse();
        assertThat(childClass.getMethod("method2").isOverridden()).isFalse();
}
```
Your implementation however gives me `childClass.getMethod("method1").isOverridden() == false`.

java-coding-prodigy committed:
Fixed the method and added a test

java-coding-prodigy commented:
I couldn't figure out a direct way to get the inherited methods of the super class I tried to use `.getMethods()` but that only returns the public methods.
If we have something like a class extending a class which also extends another class, and the lowest class has a method that the highest class has, but the middle class does not, then `.isOverridden()` returns false.
I don't know whether our API or the reflection API has an easy way to fix this or not

java-coding-prodigy committed:
Merge branch 'main' into main

hankem reviewed (changes requested):
You may have seen that the PR's `DCO` check fails. Can you please squash your changes and sign it off according to the [DCO](https://github.com/TNG/ArchUnit/blob/main/DCO)? 

It's also preferable to rebase to the current `main` instead of merging it in order to create a simple linear history.

java-coding-prodigy commented on archunit/src/test/java/com/tngtech/archunit/core/domain/JavaMethodTest.java:
```diff
@@ -0,0 +1,39 @@
+        JavaMethod childMethod1 = childClass.getMethod("method1");
+        JavaMethod baseMethod1 = baseClass.getMethod("method1");
+        assertNotEquals(childMethod1, baseMethod1);
+        assertEquals(childMethod1.getName(), baseMethod1.getName());
```
Oops, I used that assertion while I was debugging why my check failed and forgot to remove it, will do so in the next commit.

java-coding-prodigy committed:
Added support for methods in grandparent class and used ArchUnit API instead of reflection

java-coding-prodigy committed:
Merge branch 'main' of github.com:TNG/ArchUnit

java-coding-prodigy committed:
Fixed issue with `isOverridden`

I was checking if the parameters are equal by using `getParameters().equals()` which did not work as `JavaParameter#equals` checks for equality of the owner(in this case, super method) as well. `JavaType#equals` only checks for equality of the actual parameter type regardless of source, and hence works.

java-coding-prodigy committed:
Added testing for overriden method in grandparent class

java-coding-prodigy committed:
Merge branch 'main' of github.com:java-coding-prodigy/ArchUnit

hankem commented on archunit/src/main/java/com/tngtech/archunit/core/domain/JavaMethod.java:
```diff
@@ -15,20 +15,20 @@
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
```
Can we please adhere to the import order specified in [`CONTRIBUTING.md`](https://github.com/TNG/ArchUnit/blob/main/CONTRIBUTING.md), i.e. keep the `java.*` imports at the top?

java-coding-prodigy replied:
Cool, I'll do so in `JavaMethodTest` as well

hankem commented on archunit/src/main/java/com/tngtech/archunit/core/domain/JavaMethod.java:
```diff
@@ -125,6 +125,14 @@ public String getDescription() {
 
+    @PublicAPI(usage = ACCESS)
+    public boolean isOverridden() {
+        return getOwner().getAllRawSuperclasses().stream()
```
If we want to conisder overridden methods from interfaces as well, we can replace this stream with
```java
Streams.concat(getOwner().getAllRawSuperclasses().stream(), getOwner().getAllRawInterfaces().stream())
```

hankem commented on archunit/src/main/java/com/tngtech/archunit/core/domain/JavaMethod.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2014-2022 TNG Technology Consulting GmbH
+ * Copyright 2014-2023 TNG Technology Consulting GmbH
```
I guess we should isolate this _"Happy new year, code base!"_ change into a separate commit (that also fixes all java files, which happens automatically when building the project).

java-coding-prodigy replied:
Yeah for some reason my IntelliJ or some other plugin automatically made this change for all the classes with this documentation, I forgot to undo the changes on this file itself but did them for all other classes. 😅

hankem reviewed (commented):
Nice; I think you're getting close! 👍

About the [DCO](https://github.com/TNG/ArchUnit/blob/main/DCO): I would do 
* `git rebase -i origin/main`
* replace `pick` with `f` for all commits after the first one and do
* `git commit --amend -s` to create the sign-off, followed by
* `git push -f`

java-coding-prodigy committed:
Undo accidental copyright change

java-coding-prodigy committed:
Check for super methods in interfaces as well

java-coding-prodigy committed:
Import fix

hankem commented on archunit/src/main/java/com/tngtech/archunit/core/domain/JavaMethod.java:
```diff
@@ -125,6 +126,14 @@ public String getDescription() {
 
+    @PublicAPI(usage = ACCESS)
+    public boolean isOverridden() {
+        return Stream.concat(getOwner().getAllRawSuperclasses().stream(), getOwner().getAllRawInterfaces().stream())
```
Can we pleae add a small test to cover the feature that overriden methods from interfaces are detected?

hankem reviewed (commented):
Could you please sign-off your commits (most easily done after rebasing & squashing all into one), as suggested before in https://github.com/TNG/ArchUnit/pull/1040#pullrequestreview-1239752587?
Without the DCO check passing, GitHub won't let us merge the PR.

hankem commented on archunit/src/test/java/com/tngtech/archunit/core/domain/JavaMethodTest.java:
```diff
@@ -0,0 +1,58 @@
+        assertThat(grandChildClass.getMethod("method3").isOverridden()).isFalse();
+        //TODO add testing for methods with generic parameters
+    }
+}
```
We should also have a test for the use case of #1047 .

java-coding-prodigy replied:
I believe that is already in place, currently we check if the parameters and name of the method are equal, and that implies compatibility of return type.
If that is not the case, then the user code will not compile in the first place

hankem replied:
I thought that we should document that this works (which it curently does, but we also want to avoid any future regression):
```java
    @Test
    public void more_specific_return_type_is_supported() {
        class Parent {
            CharSequence method() { return "base"; }
        }
        class Child extends Parent {
            @Override
            String method() { return "overridden"; }
        }
        JavaClass childClass = new ClassFileImporter().importClass(Child.class);
        JavaMethod method = childClass.getMethod("method");
        assertThat(method.isOverridden()).isTrue();
    }
```

java-coding-prodigy committed:
Merge branch 'main' into main

hankem commented on archunit/src/test/java/com/tngtech/archunit/core/domain/JavaMethodTest.java:
```diff
@@ -0,0 +1,58 @@
+        assertThat(grandChildClass.getMethod("method1", int.class).isOverridden()).isTrue();
+        assertThat(grandChildClass.getMethod("method2").isOverridden()).isTrue();
+        assertThat(grandChildClass.getMethod("method3").isOverridden()).isFalse();
+        //TODO add testing for methods with generic parameters
```
It seems that we indeed don't recognize all overriden generic methods:
```java
    @Test
    public void overridden_generic_methods_are_supported() {
        class Parent<T extends Number> {
            void method(T t) { }
        }
        class Child extends Parent<Integer> {
            @Override
            void method(Integer t) { }
        }
        JavaClass childClass = new ClassFileImporter().importClass(Child.class);
        JavaMethod method = childClass.getMethod("method", Integer.class);
        assertThat(method.isOverridden()).isTrue();  // Expecting value to be true but was false 
    }
```

java-coding-prodigy replied:
Hmmm, this might complicate the code now.
I'll see what I can do.

java-coding-prodigy committed:
Got this so far

java-coding-prodigy committed:
Got this so far

java-coding-prodigy commented:
So far I think I know what we can do to fix the issue.
We need to get the supperclasses and interfaces of the method's class but they cannot be the raw ones as they don't have type arguments.
So looks like we can keep on calling a .`supperclass()` again and again unti we reach `java.lang.Object` as well as use `.interfaces`
However this solves the current problem at hand, it may not be bug-free.

Also ignore `JavaMethodTest.java` that is just me playing around with the objects while debugging.

java-coding-prodigy committed:
Merge branch 'main' into main

java-coding-prodigy committed:
Merge branch 'main' into main

java-coding-prodigy committed:
Merge branch 'main' into main

codecholeric commented:
@java-coding-prodigy are you still working on this? Do you need any support?

java-coding-prodigy commented:
> @java-coding-prodigy are you still working on this? Do you need any support?

I was working on it some time ago and I could continue now. However I had run into some issues earlier.
I'll see what I can do.

java-coding-prodigy committed:
Merge branch 'TNG:main' into main

java-coding-prodigy committed:
Finally passed the test.

java-coding-prodigy committed:
Merge branch 'main' into main

java-coding-prodigy commented:
Fixed the generic issue, however code is too ugly and slow, we should try to enhance it.
I have only done this for classes, but I presume it would be similar for interfaces too.

java-coding-prodigy commented:
I'm thinking of adding a `JavaMethod#overridenFrom(JavaType` and make `isOverriden` call that method. What do you think @hankem  ?
